istio/src/main/scala/io/buoyant/k8s/istio/mixer/MixerClient.scala

```diff
@@ -0,0 +1,118 @@
+    case _: ResourceExhausted => 429
+    case _: FailedPrecondition => 400
+    case _: Aborted => 409
+    case _: OutOfRange => 400
+    case _: Unimplemented => 501
+    case _: Internal => 500
+    case _: Unavailable => 503
+    case _: DataLoss => 500
+    case _: Unauthenticated => 401
+    case _: Other => 500
+  }
+}
+
+class MixerClient(client: Mixer) {
+  private[this] val log = Logger(this.getClass.getSimpleName)
+  private[this] val codeUsedWhenUnknownStatus = Unknown("NO RESPONSE FROM SERVER")
+
+  /**
+   * Sends a report to Istio' Mixer, using the [[https://istio.io/docs/reference/api/mixer/mixer-service.html#report Report API]]
+   *
+   * minimum set of attributes to generate the following metrics in mixer/prometheus:
+   * - request_count
+   * - request_duration_bucket
+   * - request_duration_count
+   * - request_duration_sum
+   * *
+   * example metrics exposed by mixer/prometheus:
+   * request_count{method="/productpage",response_code="200",service="productpage",source="unknown",target="productpage.default.svc.cluster.local",version="v1"} 327
+   * request_count{method="/reviews",response_code="500",service="reviews",source="productpage",target="reviews.default.svc.cluster.local",version="v1"} 1
+   */
+  def report(
+    responseCode: ResponseCodeIstioAttribute,
+    requestPath: RequestPathIstioAttribute,
+    targetService: TargetServiceIstioAttribute,
+    sourceLabel: SourceLabelIstioAttribute,
+    targetLabel: TargetLabelsIstioAttribute,
+    duration: ResponseDurationIstioAttribute
+  ): Future[Unit] = {
+
+    val reportRequest = MixerApiRequests.mkReportRequest(
+      responseCode,
+      requestPath,
+      targetService,
+      sourceLabel,
+      targetLabel,
+      duration
+    )
+    log.trace("MixerClient.report: %s", reportRequest)
+    client.report(Stream.value(reportRequest))
+    Future.Done
+  }
+
+  /**
+   * Checks this request agains  ixer' Pre-Condition system, using the [[https://istio.io/docs/reference/api/mixer/mixer-service.html#check Check API]]
+   * @param istioRequest
+   * @return
+   */
+  def checkPreconditions(istioRequest: IstioRequest[_]): Future[MixerCheckStatus] = {
+    val checkRequest = MixerApiRequests.mkCheckRequest(istioRequest)
+    log.trace("MixerClient.check: %s", checkRequest)
+
+    client.check(Stream.value(checkRequest)).recv().transform {
+      case Return(stream) => mkStatus(stream)
+      case Throw(e) => handleExceptionInPreconditionCheck(e)
+    }
+  }
+
+  private def handleExceptionInPreconditionCheck(e: Throwable) = {
+    //TODO: Consider making this behaviour configurable, as per https://github.com/istio/mixerclient/blob/13460a96b4b3aa792ad47817d4fbf529b63c25e2/src/check_cache.cc#L163
+    log.error("Error checking Mixer pre-conditions: %s - %s", e.getClass, e.getMessage);
+    Future.value(MixerCheckStatus(codeUsedWhenUnknownStatus))
+  }
+
+  private[this] def mkStatus(stream: Stream.Releasable[CheckResponse]) = {
```

Comment:
I think a `Stream.Releasable` is an item from the stream, not the stream itself.

Reply:
Fixed, thanks!